lib/drip/client.rb

```diff
@@ -68,69 +68,31 @@ def content_type
       JSON_API_CONTENT_TYPE
     end
 
-    def get(url, options = {})
-      make_request(Net::HTTP::Get, make_uri(url), options)
-    end
-
-    def post(url, options = {})
-      make_request(Net::HTTP::Post, make_uri(url), options)
+    Drip::Request::VERB_CLASS_MAPPING.keys.each do |verb|
+      define_method(verb) do |path, options = {}|
+        warn "[DEPRECATED] Drip::Client##{verb} please use the API endpoint specific methods"
+        make_v2_request(verb, "v2/#{path}", options)
+      end
     end
 
-    def put(url, options = {})
-      make_request(Net::HTTP::Put, make_uri(url), options)
-    end
+  private
 
-    def delete(url, options = {})
-      make_request(Net::HTTP::Delete, make_uri(url), options)
+    def make_v2_request(http_verb, path, options = {})
+      make_request Drip::Request.new(http_verb, make_uri(path), options, JSON_API_CONTENT_TYPE)
     end
 
-  private
-
     def private_generate_resource(key, *args)
       # No reason for this to be part of the public API, so making a duplicate method to make it private.
       { key => args }
     end
 
     def make_uri(path)
-      if !path.start_with?("v2/") && !path.start_with?("v3/")
-        warn "[DEPRECATED] Automatically prepended path with 'v2/'"
-        path = "v2/#{path}"
-      end
       URI(@config.url_prefix) + URI(path)
     end
 
-    def make_request(verb_klass, uri, options, step = 0)
-      raise TooManyRedirectsError, 'too many HTTP redirects' if step >= REDIRECT_LIMIT
-
+    def make_request(drip_request)
       build_response do
-        Net::HTTP.start(uri.host, uri.port, connection_options(uri.scheme)) do |http|
-          if verb_klass == Net::HTTP::Get
-            uri.query = URI.encode_www_form(options)
-          end
-
-          request = verb_klass.new uri
-
-          unless verb_klass == Net::HTTP::Get
-            request.body = options.to_json
-          end
-
-          request['User-Agent'] = "Drip Ruby v#{Drip::VERSION}"
-          request['Content-Type'] = JSON_API_CONTENT_TYPE
-          request['Accept'] = "*/*"
-
-          if @config.access_token
-            request['Authorization'] = "Bearer #{@config.access_token}"
-          else
-            request.basic_auth @config.api_key, ""
-          end
-
-          response = http.request request
-          if response.is_a?(Net::HTTPRedirection)
-            return make_request(verb_klass, URI(response["Location"]), options, step + 1)
-          else
-            response
-          end
-        end
+        Drip::Client::HTTPClient.new(@config).make_request(drip_request)
```

Comment:
I'd suggest to create an instance of `Drip::Client::HTTPClient` once, memoize it and reset if config is changed.

Reply:
What's the benefit of doing so? If we're going to be introducing some cache invalidation issues, I'd like to have a good reason.

Reply:
The point is you don't need to create a new instance of http client for each request. It leads to memory leaks and creates performance overhead. This object depends only on `@config` so I think it should be quite easy to reset it when config is changed.